### ISSUE TITLE
The new transaction autocomplete behaviour in develop may have usability problems

### ISSUE THREAD

**Issue by rivaldi8** [codinguser/gnucash-android#309]
I've had some confusion with the new transaction autocomplete function based on templates and I think it make bring some usability problems.

I've been looking for advantages of the new system, but the only I can think of is that now the user can select which transactions will appear in the autocomplete llist. This makes the list smaller and saves typing.

However, I see some disadvantages:

* When the user upgrades she'll have to either edit each transaction and check the option for it to appear in the dropdown or she'll have to type again these transactions.
* Requires more interaction from the user. In addition, she now has to scroll to the end of the form and check the option.
* The user must know the option exists. Taking into account that it's at the end of the form and, provably below the fold, makes it more difficult for the user to discover it.
* The user may not relate the "Save as template" with the autocompletion. In my case I didn't relate it until I looked at the code.
* The user may forget to check the "Save as template" option. He'd have to edit the transaction again to check the option or remember to check it when he enters the transaction again the next time.
* The behaviour is different from that of other applications and from previous versions of GnuCash. This can confuse users that expect the previous behaviour [[1]](http://www.nngroup.com/articles/the-top-ten-web-design-mistakes-of-1999/)(point 3) and think (like me) that it's a bug. I can't think of any application where the autocomplete entries of a form are explicitly selected by the user.
* The option adds more clutter to the interface. In my opinion, GnuCash should strive to have simpler screens instead of adding more options.
* If the user decides he no longer wants the transaction to appear in the dropdown, it may be difficult for him to find the one he marked as template.

Is there any other problem that the new behaviour solves apart from the number of transactions in the autocomplete dropdown? I there isn't, I'd rather go for improving the quality of the entries. For example, listing at the top the ones matching from the same account, the more recent, avoiding duplicates, etc.

Right now, in the worst case, the user just has to continue typing some more to filter out the unwanted entries.

**Comment by monoman** [comment]
In truth, I like the concept of the change, but the real nice thing it
would allow is to have a new widget to jump directly into adding a
transaction with an specific template, and then after that notifications
coming predicting transactions, mostly expenses, that are regular (like
lunch meals) but may vary on amount, and allowing to add from the
notifications if the amount is the same...

Just presenting the idea. Later I'll try to open specific issues and maybe
send some PRs.

Rafael Teixeira
O..:.)oooo

**Comment by codinguser** [comment]
@rivaldi8 Thanks for the feedback. 
I agree with you that there is utility in providing auto-complete with all the transactions in the database (that's why I put the feature there in the first place). 
Maybe it could be eventually improved with some of the enhancements you mention.

That said, there are a few things to consider:
* Firstly, transaction templates have a lot more uses than just for auto-complete, e.g. scheduled actions, the widget (as @monoman mentioned) etc. So they're here to stay.
* Also, note that the templates feature as it is now in the develop branch is not complete and of course there will be UI for users to edit and/or delete templates.
* I disagree that this is fundamentally different from the expected use of an auto-complete box. Yes, it may break the expectations of existing users but I've had more complaints than positive feedback about the feature as it is now.
* I also disagree that it clutters the interface. Otherwise everything since amount & description fields can be considered clutter as well. I don't think so.

Development is ongoing so we might revisit the auto-complete feature again if we find ways to improve what it is now. Thanks for highlighting this issue!

**Comment by codinguser** [comment]
Updated code to include both templates and existing transactions, while improving the selection and display of suggestions. Let me know what you think @rivaldi8

**Comment by rivaldi8** [comment]
That was fast! I like it very much the way the autocomplete appears now. In my opinion improving the autocomplete suggestions is the way to go. Most of the time I enter transactions one-handed while walking on the street, so any additional interaction is a big pain for me.

The widget mentioned by @monoman was the way I expected the templates feature to work. That's because I got confused when the autocompletion stopped working.

By the way, sorry if I've complained too early on an unfinished feature. I'll better wait for the beta the next time :)
